File: apps/nestjs-api/src/modules/analytics/analytics-plot.ts
import { sql, type SQL } from "drizzle-orm";
import type { Database } from "../../infra/database/drizzle.module";

// Port of plane/utils/analytics_plot.py — VALID_* + build_graph_plot.
export const VALID_ANALYTICS_FIELDS = [
  "state_id",
  "state__group",
  "labels__id",
  "assignees__id",
  "estimate_point__value",
  "issue_cycle__cycle_id",
  "issue_module__module_id",
  "priority",
  "start_date",
  "target_date",
  "created_at",
  "completed_at",
] as const;

export const VALID_YAXIS = ["issue_count", "estimate"] as const;

const DATE_FIELDS = new Set(["created_at", "start_date", "target_date", "completed_at"]);

interface Resolved {
  expr: SQL;
  join?: { alias: string; sql: SQL };
  isDate: boolean;
}

function resolveField(field: string): Resolved {
  switch (field) {
    case "state_id":
      return { expr: sql`i.state_id::text`, isDate: false };
    case "state__group":
      return { expr: sql`sg."group"`, join: { alias: "sg", sql: sql`LEFT JOIN states sg ON sg.id = i.state_id` }, isDate: false };
    case "labels__id":
      return {
        expr: sql`il.label_id::text`,
        join: { alias: "il", sql: sql`LEFT JOIN issue_labels il ON il.issue_id = i.id AND il.deleted_at IS NULL` },
        isDate: false,
      };
    case "assignees__id":
      return {
        expr: sql`ia.assignee_id::text`,
        join: { alias: "ia", sql: sql`LEFT JOIN issue_assignees ia ON ia.issue_id = i.id AND ia.deleted_at IS NULL` },
        isDate: false,
      };
    case "estimate_point__value":
      return { expr: sql`ep.value`, join: { alias: "ep", sql: sql`LEFT JOIN estimate_points ep ON ep.id = i.estimate_point_id` }, isDate: false };
    case "issue_cycle__cycle_id":
      return {
        expr: sql`ci.cycle_id::text`,
        join: { alias: "ci", sql: sql`LEFT JOIN cycle_issues ci ON ci.issue_id = i.id AND ci.deleted_at IS NULL` },
        isDate: false,
      };
    case "issue_module__module_id":
      return {
        expr: sql`mi.module_id::text`,
        join: { alias: "mi", sql: sql`LEFT JOIN module_issues mi ON mi.issue_id = i.id AND mi.deleted_at IS NULL` },
        isDate: false,
      };
    case "priority":
      return { expr: sql`i.priority`, isDate: false };
    case "start_date":
    case "target_date":
    case "created_at":
    case "completed_at": {
      const col = sql.raw(`i.${field}`);
      // Django Concat(ExtractYear, '-', ExtractMonth) — no zero-padding ("2024-3").
      return { expr: sql`(EXTRACT(YEAR FROM ${col})::int || '-' || EXTRACT(MONTH FROM ${col})::int)`, isDate: true };
    }
    default:
      throw new Error(`Invalid analytics field: ${field}`);
  }
}

// sort_data(): priority order, else alpha with lowercase "none" last.
function sortData(data: Record<string, unknown[]>, tempAxis: string): Record<string, unknown[]> {
  const out: Record<string, unknown[]> = {};
  if (tempAxis === "priority") {
    for (const key of ["low", "medium", "high", "urgent", "none"]) {
      if (key in data) out[key] = data[key];
    }
    return out;
  }
  const keys = Object.keys(data).sort((a, b) => {
    const an = a === "none" ? 1 : 0;
    const bn = b === "none" ? 1 : 0;
    if (an !== bn) return an - bn;
    return a < b ? -1 : a > b ? 1 : 0;
  });
  for (const k of keys) out[k] = data[k];
  return out;
}

export interface GraphPlotArgs {
  where: SQL[]; // base conditions (workspace + issue_objects + issue_filters) over alias `i`
  xAxis: string;
  yAxis: string;
  segment?: string | null;
}

/**
 * Faithful port of build_graph_plot: group the (filtered) issue set by an x-axis dimension (+ optional
 * segment), counting issues or summing estimate values, returning { [dimension]: [rows...] } sorted.
 */
export async function buildGraphPlot(db: Database, { where, xAxis, yAxis, segment }: GraphPlotArgs): Promise<Record<string, unknown[]>> {
  if (!(VALID_ANALYTICS_FIELDS as readonly string[]).includes(xAxis)) throw new Error(`Invalid x_axis value: ${xAxis}`);
  if (!(VALID_YAXIS as readonly string[]).includes(yAxis)) throw new Error(`Invalid y_axis value: ${yAxis}`);
  if (segment && !(VALID_ANALYTICS_FIELDS as readonly string[]).includes(segment)) throw new Error(`Invalid segment value: ${segment}`);

  const x = resolveField(xAxis);
  const seg = segment ? resolveField(segment) : null;

  // Collect joins, deduped by alias. Estimate y-axis always needs estimate_points.
  const joins = new Map<string, SQL>();
  if (x.join) joins.set(x.join.alias, x.join.sql);
  if (seg?.join) joins.set(seg.join.alias, seg.join.sql);
  if (yAxis === "estimate" && !joins.has("ep")) joins.set("ep", sql`LEFT JOIN estimate_points ep ON ep.id = i.estimate_point_id`);

  const conds = [...where];
  if (x.isDate) conds.push(sql`${x.expr} IS NOT NULL`);

  const selectParts: SQL[] = [sql`${x.expr} AS dimension`];
  if (seg) selectParts.push(sql`${seg.expr} AS segment`);
  const agg = yAxis === "issue_count" ? sql`COUNT(*) AS count` : sql`SUM(CAST(ep.value AS double precision)) AS estimate`;
  selectParts.push(agg);

  const groupBy: SQL[] = [sql`dimension`];
  if (seg) groupBy.push(sql`segment`);

  const joinSql = joins.size ? sql.join([...joins.values()], sql` `) : sql``;
  const whereSql = sql.join(conds, sql` AND `);

  const query = sql`SELECT ${sql.join(selectParts, sql`, `)} FROM issues i ${joinSql} WHERE ${whereSql} GROUP BY ${sql.join(groupBy, sql`, `)} ORDER BY dimension`;

  const res = await db.execute(query);
  const rows = (res as unknown as { rows: Array<Record<string, unknown>> }).rows;

  // groupby dimension → { str(dimension): [rows] }, then sort_data.
  const grouped: Record<string, unknown[]> = {};
  for (const r of rows) {
    const key = r.dimension == null ? "None" : String(r.dimension);
    const row: Record<string, unknown> = { dimension: r.dimension == null ? null : String(r.dimension) };
    if (seg) row.segment = r.segment ?? null;
    if (yAxis === "issue_count") row.count = Number(r.count);
    else row.estimate = r.estimate == null ? null : Number(r.estimate);
    (grouped[key] ??= []).push(row);
  }
  return sortData(grouped, xAxis);
}

File: apps/api/plane/utils/analytics_plot.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

"""Query helpers that build analytics and burndown chart data from Issue querysets.

build_graph_plot powers the analytics views (issue count or estimate sum grouped by an
x-axis field and optional segment); burndown_plot computes the day-by-day remaining work
for a cycle or module.
"""

# Python imports
from datetime import timedelta
from itertools import groupby

# Django import
from django.db import models
from django.db.models import Case, CharField, Count, F, Sum, Value, When, FloatField
from django.db.models.functions import (
    Coalesce,
    Concat,
    ExtractMonth,
    ExtractYear,
    TruncDate,
    Cast,
)
from django.utils import timezone

# Module imports
from plane.db.models import Issue, Project

# Whitelist of Issue lookups allowed as x-axis or segment (guards against arbitrary field access).
VALID_ANALYTICS_FIELDS = [
    "state_id",
    "state__group",
    "labels__id",
    "assignees__id",
    "estimate_point__value",
    "issue_cycle__cycle_id",
    "issue_module__module_id",
    "priority",
    "start_date",
    "target_date",
    "created_at",
    "completed_at",
]

# Supported metrics: number of issues, or sum of estimate point values.
VALID_YAXIS = ["issue_count", "estimate"]


def annotate_with_monthly_dimension(queryset, field_name, attribute):
    """Annotate `attribute` with a "YYYY-M" string built from the date field `field_name`."""
    # Get the year and the months
    year = ExtractYear(field_name)
    month = ExtractMonth(field_name)
    # Concat the year and month
    dimension = Concat(year, Value("-"), month, output_field=CharField())
    # Annotate the dimension
    return queryset.annotate(**{attribute: dimension})


def extract_axis(queryset, x_axis):
    """Annotate the queryset with a `dimension` column for the given x-axis.

    Date fields are bucketed by year-month; other fields are copied as-is.
    Returns (queryset, "dimension"). Raises ValueError for non-whitelisted fields.
    """
    if x_axis not in VALID_ANALYTICS_FIELDS:
        raise ValueError(f"Invalid x_axis value: {x_axis}")
    # Format the dimension when the axis is in date
    if x_axis in ["created_at", "start_date", "target_date", "completed_at"]:
        queryset = annotate_with_monthly_dimension(queryset, x_axis, "dimension")
        return queryset, "dimension"
    else:
        return queryset.annotate(dimension=F(x_axis)), "dimension"


def sort_data(data, temp_axis):
    """Order grouped chart data: fixed low->urgent order for priority, else by key with "none" last."""
    # When the axis is in priority order by
    if temp_axis == "priority":
        order = ["low", "medium", "high", "urgent", "none"]
        return {key: data[key] for key in order if key in data}
    else:
        return dict(sorted(data.items(), key=lambda x: (x[0] == "none", x[0])))


def build_graph_plot(queryset, x_axis, y_axis, segment=None):
    """Return analytics data grouped by x-axis value, optionally split by a segment field.

    Result is {dimension: [row, ...]} where each row holds the dimension, optional segment and
    either `count` (y_axis="issue_count") or `estimate` (y_axis="estimate").
    Raises ValueError for invalid axis/segment names.
    """
    if x_axis not in VALID_ANALYTICS_FIELDS:
        raise ValueError(f"Invalid x_axis value: {x_axis}")
    if y_axis not in VALID_YAXIS:
        raise ValueError(f"Invalid y_axis value: {y_axis}")
    if segment and segment not in VALID_ANALYTICS_FIELDS:
        raise ValueError(f"Invalid segment value: {segment}")

    # temp x_axis
    temp_axis = x_axis
    # Extract the x_axis and queryset
    queryset, x_axis = extract_axis(queryset, x_axis)
    if x_axis == "dimension":
        queryset = queryset.exclude(dimension__isnull=True)

    # Date segments are bucketed by year-month, like date x-axes.
    if segment in ["created_at", "start_date", "target_date", "completed_at"]:
        queryset = annotate_with_monthly_dimension(queryset, segment, "segmented")
        segment = "segmented"

    queryset = queryset.values(x_axis)

    # Issue count
    if y_axis == "issue_count":
        # NOTE: is_null / dimension_ex are dropped by the chained .values("dimension") below.
        queryset = queryset.annotate(
            is_null=Case(
                When(dimension__isnull=True, then=Value("None")),
                default=Value("not_null"),
                output_field=models.CharField(max_length=8),
            ),
            dimension_ex=Coalesce("dimension", Value("null")),
        ).values("dimension")
        queryset = queryset.annotate(segment=F(segment)) if segment else queryset
        queryset = queryset.values("dimension", "segment") if segment else queryset.values("dimension")
        queryset = queryset.annotate(count=Count("*")).order_by("dimension")

    # Estimate
    else:
        queryset = queryset.annotate(estimate=Sum(Cast("estimate_point__value", FloatField()))).order_by(x_axis)
        queryset = queryset.annotate(segment=F(segment)) if segment else queryset
        queryset = (
            queryset.values("dimension", "segment", "estimate") if segment else queryset.values("dimension", "estimate")
        )

    # itertools.groupby only merges adjacent rows, so this relies on the order_by above.
    result_values = list(queryset)
    grouped_data = {str(key): list(items) for key, items in groupby(result_values, key=lambda x: x[str("dimension")])}

    return sort_data(grouped_data, temp_axis)


def burndown_plot(queryset, slug, project_id, plot_type, cycle_id=None, module_id=None):
    """Compute burndown chart data for a cycle (cycle_id) or module (module_id).

    `queryset` is the Cycle/Module instance (annotated with total_issues). plot_type "points"
    burns down estimate points (only when the project uses a points estimate), otherwise
    issue counts. Returns {date_str: remaining}; future dates map to None.
    """
    # Total Issues in Cycle or Module
    total_issues = queryset.total_issues
    # check whether the estimate is a point or not
    estimate_type = Project.objects.filter(
        workspace__slug=slug,
        pk=project_id,
        estimate__isnull=False,
        estimate__type="points",
    ).exists()
    if estimate_type and plot_type == "points" and cycle_id:
        issue_estimates = Issue.issue_objects.filter(
            workspace__slug=slug,
            project_id=project_id,
            issue_cycle__cycle_id=cycle_id,
            issue_cycle__deleted_at__isnull=True,
            estimate_point__isnull=False,
        ).values_list("estimate_point__value", flat=True)

        issue_estimates = [float(value) for value in issue_estimates]
        total_estimate_points = sum(issue_estimates)

    if estimate_type and plot_type == "points" and module_id:
        issue_estimates = Issue.issue_objects.filter(
            workspace__slug=slug,
            project_id=project_id,
            issue_module__module_id=module_id,
            issue_module__deleted_at__isnull=True,
            estimate_point__isnull=False,
        ).values_list("estimate_point__value", flat=True)

        issue_estimates = [float(value) for value in issue_estimates]
        total_estimate_points = sum(issue_estimates)

    if cycle_id:
        if queryset.end_date and queryset.start_date:
            # Get all dates between the two dates
            date_range = [
                (queryset.start_date + timedelta(days=x)).date()
                for x in range((queryset.end_date.date() - queryset.start_date.date()).days + 1)
            ]
        else:
            date_range = []

        # Seed every day in the range with 0 so days without completions still appear.
        chart_data = {str(date): 0 for date in date_range}

        if plot_type == "points":
            completed_issues_estimate_point_distribution = (
                Issue.issue_objects.filter(
                    workspace__slug=slug,
                    project_id=project_id,
                    issue_cycle__cycle_id=cycle_id,
                    issue_cycle__deleted_at__isnull=True,
                    estimate_point__isnull=False,
                )
                .annotate(date=TruncDate("completed_at"))
                .values("date")
                .values("date", "estimate_point__value")
                .order_by("date")
            )
        else:
            completed_issues_distribution = (
                Issue.issue_objects.filter(
                    workspace__slug=slug,
                    project_id=project_id,
                    issue_cycle__cycle_id=cycle_id,
                    issue_cycle__deleted_at__isnull=True,
                )
                .annotate(date=TruncDate("completed_at"))
                .values("date")
                .annotate(total_completed=Count("id"))
                .values("date", "total_completed")
                .order_by("date")
            )

    if module_id:
        # Get all dates between the two dates
        date_range = [
            (queryset.start_date + timedelta(days=x))
            for x in range((queryset.target_date - queryset.start_date).days + 1)
        ]

        chart_data = {str(date): 0 for date in date_range}

        if plot_type == "points":
            completed_issues_estimate_point_distribution = (
                Issue.issue_objects.filter(
                    workspace__slug=slug,
                    project_id=project_id,
                    issue_module__module_id=module_id,
                    issue_module__deleted_at__isnull=True,
                    estimate_point__isnull=False,
                )
                .annotate(date=TruncDate("completed_at"))
                .values("date")
                .values("date", "estimate_point__value")
                .order_by("date")
            )
        else:
            completed_issues_distribution = (
                Issue.issue_objects.filter(
                    workspace__slug=slug,
                    project_id=project_id,
                    issue_module__module_id=module_id,
                    issue_module__deleted_at__isnull=True,
                )
                .annotate(date=TruncDate("completed_at"))
                .values("date")
                .annotate(total_completed=Count("id"))
                .values("date", "total_completed")
                .order_by("date")
            )

    # Remaining work for each day = total minus everything completed on or before that day.
    if plot_type == "points":
        for date in date_range:
            # NOTE: total_estimate_points is only set when the project uses a points estimate.
            cumulative_pending_issues = total_estimate_points
            total_completed = 0
            total_completed = sum(
                float(item["estimate_point__value"])
                for item in completed_issues_estimate_point_distribution
                if item["date"] is not None and item["date"] <= date
            )
            cumulative_pending_issues -= total_completed
            if date > timezone.now().date():
                chart_data[str(date)] = None
            else:
                chart_data[str(date)] = cumulative_pending_issues
    else:
        for date in date_range:
            cumulative_pending_issues = total_issues
            total_completed = 0
            total_completed = sum(
                item["total_completed"]
                for item in completed_issues_distribution
                if item["date"] is not None and item["date"] <= date
            )
            cumulative_pending_issues -= total_completed
            if date > timezone.now().date():
                chart_data[str(date)] = None
            else:
                chart_data[str(date)] = cumulative_pending_issues

    return chart_data
